kai/service/incident_store/psql.py

```diff
@@ -1,93 +1,182 @@
 import argparse
 import datetime
-import json
+import enum
 import os
-from functools import wraps
-from inspect import signature
+from typing import Any, Optional
 from urllib.parse import unquote, urlparse
 
-import psycopg2
 from git import Repo
-from psycopg2.extensions import connection
-from psycopg2.extras import DictCursor, DictRow
+from sqlalchemy import (
+    Column,
+    DateTime,
+    ForeignKey,
+    ForeignKeyConstraint,
+    String,
+    create_engine,
+    func,
+    select,
+)
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.orm import DeclarativeBase, Mapped, Session, mapped_column, relationship
+from sqlalchemy.types import JSON
 
-from kai.constants import PATH_SQL
-from kai.embedding_provider import EmbeddingNone
 from kai.kai_logging import KAI_LOG
 from kai.models.kai_config import KaiConfig, KaiConfigIncidentStorePostgreSQLArgs
 from kai.report import Report
 from kai.service.incident_store.incident_store import (
     Application,
     IncidentStore,
     Solution,
-    filter_incident_vars,
     load_reports_from_directory,
-    remove_known_prefixes,
 )
 
+# have a logical category of un-mutated reports and one with the mutations
 
-def supply_cursor_if_none(func):
-    @wraps(func)
-    def wrapper(self, *args, **kwargs):
-        sig = signature(func)
-        bound_args = sig.bind(self, *args, **kwargs)
-        bound_args.apply_defaults()
 
-        if "cur" not in bound_args.arguments or bound_args.arguments["cur"] is None:
-            with self.conn.cursor() as cur:
-                bound_args.arguments["cur"] = cur
-                return func(*bound_args.args, **bound_args.kwargs)
-        else:
-            return func(*bound_args.args, **bound_args.kwargs)
+class Base(DeclarativeBase):
+    type_annotation_map = {
+        dict[str, Any]: JSON().with_variant(JSONB(), "postgresql"),
+        list[str]: JSON().with_variant(JSONB(), "postgresql"),
+    }
 
-    return wrapper
 
+class PSQLUnmodifiedReport(Base):
+    __tablename__ = "unmodified_reports"
 
-# TODO(@JonahSussman): Migrate this to use an ORM
-class PSQLIncidentStore(IncidentStore):
-    def __init__(self, args: KaiConfigIncidentStorePostgreSQLArgs):
-        self.emb_provider = EmbeddingNone()
+    application_name: Mapped[str] = mapped_column(primary_key=True)
+    generated_at: Mapped[datetime.datetime] = mapped_column(
```

Comment:
I think we probably don't want to use this as a primary key, since it will be unique if you insert the same report twice. I think ideally if we picked up the same report twice we would not duplicate it.

Reply:
My thought process was that this was just a record of all reports in the order they were processed. What are you thinking as the primary key?

Reply:
I think that is accurate, but we also don't want to duplicate the reports coming in. I think ideally we can pull an identifier or combination of identifiers off the Report itself. I say this mostly because the Hub importer polls the API, but if it restarts it could suck up all the previous reports again, so we just want to make sure those only go in once (or we could see like a `crashloop but successfully load 50 reports` where the database just balloons in size). We could do that filter elsewhere as well if needed but if there's an identifier on the report that can be used I think that would be ideal

Reply:
Gotcha, that makes sense. How about storing the commit that generated the report instead?

Reply:
hmm, I think that could be part of an identifier but we could feasibly get multiple reports against the same commit. I'd be cool adding an ID field to the report object and requiring us to pass it in. I think we'll have report IDs on the konveyor side so we'd just need to figure out how to populate it in the on-disk scenario

Reply:
Ok solid, I think I see what you're saying now